pkg/i2gw/providers/openapi3/converter.go

```diff
@@ -0,0 +1,619 @@
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package openapi3
+
+import (
+	"fmt"
+	"log"
+	"regexp"
+	"slices"
+	"sort"
+	"strconv"
+	"strings"
+
+	"github.com/getkin/kin-openapi/openapi3"
+	"github.com/samber/lo"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/runtime/schema"
+	"k8s.io/apimachinery/pkg/types"
+	"k8s.io/apimachinery/pkg/util/validation/field"
+	gatewayv1 "sigs.k8s.io/gateway-api/apis/v1"
+	gatewayv1beta1 "sigs.k8s.io/gateway-api/apis/v1beta1"
+
+	"github.com/kubernetes-sigs/ingress2gateway/pkg/i2gw"
+	"github.com/kubernetes-sigs/ingress2gateway/pkg/i2gw/providers/common"
+)
+
+const (
+	HostWildcard   = "*"
+	HostSeparator  = ","
+	ParamSeparator = ","
+
+	HTTPRouteRulesMax      = 16
+	HTTPRouteMatchesMax    = 8
+	HTTPRouteMatchesMaxMax = HTTPRouteRulesMax * HTTPRouteMatchesMax
+)
+
+var uriRegexp = regexp.MustCompile(`^((https?)://([^/]+))?(/.*)?$`) // [_][1] = scheme, [_][2] = host, [_][3] = path
+
+type Converter interface {
+	Convert(Storage) (i2gw.GatewayResources, field.ErrorList)
+}
+
+// NewConverter returns a converter of OpenAPI Specifications 3.x from a storage into Gateway API resources.
+func NewConverter(conf *i2gw.ProviderConf) Converter {
+	converter := &converter{
+		namespace:    conf.Namespace,
+		tlsSecretRef: types.NamespacedName{},
+		backendRef:   toBackendRef(""),
+	}
+
+	if ps := conf.ProviderSpecific[ProviderName]; ps != nil {
+		converter.gatewayClassName = ps[GatewayClassFlag]
+		converter.tlsSecretRef = toNamespacedName(ps[TlsSecretFlag])
+		converter.backendRef = toBackendRef(ps[BackendFlag])
+	}
+
+	return converter
+}
+
+type backendRef struct {
+	types.NamespacedName
+	port *gatewayv1.PortNumber
+}
+
+type converter struct {
+	namespace        string
+	gatewayClassName string
+	tlsSecretRef     types.NamespacedName
+	backendRef       backendRef
+}
+
+var _ Converter = &converter{}
+
+func (c *converter) Convert(storage Storage) (i2gw.GatewayResources, field.ErrorList) {
+	gatewayResources := i2gw.GatewayResources{
+		Gateways:        make(map[types.NamespacedName]gatewayv1.Gateway),
+		HTTPRoutes:      make(map[types.NamespacedName]gatewayv1.HTTPRoute),
+		ReferenceGrants: make(map[types.NamespacedName]gatewayv1beta1.ReferenceGrant),
+	}
+
+	var errors field.ErrorList
+	resourcesNamePrefixes := make(map[string]int)
+
+	for _, spec := range storage.GetResources() {
+		resourcesNamePrefix := toResourcesNamePrefix(spec)
+		if _, exists := resourcesNamePrefixes[resourcesNamePrefix]; !exists {
+			resourcesNamePrefixes[resourcesNamePrefix] = 0
+		}
+		resourcesNamePrefixes[resourcesNamePrefix]++
+		if resourcesNamePrefixes[resourcesNamePrefix] > 1 {
+			resourcesNamePrefix = fmt.Sprintf("%s-%d", resourcesNamePrefix, resourcesNamePrefixes[resourcesNamePrefix]+1)
```

Comment:
can we add logs and/or comments here to explain the logic?

It will also be useful for @Devaansh-Kumar project to build notifications system for conversion results.

Reply:
similar comments for the the helper functions below, adding a line explaining when we you take a not obvious decisions will be very useful

Reply:
Sure. I can add the comment.

Out of curiosity though, what would a more obvious decision here be in your opinion?

Reply:
I've added comments. Please let me know if we want logs as well, in which case I'd probably also expect some definition regarding https://github.com/kubernetes-sigs/ingress2gateway/pull/157#discussion_r1608602572, to avoid multiple formats of log messages.

Reply:
sorry, I did not meant that your choice was not obvious as "I would choose something else". But when a user is invoking the tool, I am not sure it will be straight forward to them. 
@Devaansh-Kumar is working on a notification system, so he might consider this as an INFO notification,  and comments in the code will help him.
Hopefully this makes sense. 

No need for logs.

Reply:
No problem, @LiorLieberman! You're absolutely right about making things clear. What's obvious to some people may not be to others indeed. I'm glad you brought it up. Thanks!